### DISCUSSION TITLE
Object Detection example - Blurred Image saved

### DISCUSSION THREAD

**Discussion by AmarRobo** [AxisCommunications/acap-native-sdk-examples#104] [General] [answered]
I am trying the object detection example [acap-native-sdk-examples].(https://github.com/AxisCommunications/acap-native-sdk-examples). The saved crop image on the device in the folder /var/volatile/ tmp/ is blurred. I tried with default quality 80.  
**![detection_0](https://user-images.githubusercontent.com/27479245/209511377-9ba84349-3c0c-4587-836e-9fcdda7f9bf9.jpg)**
I tried saving a 640 X 640 from the raw stream using jpeg buffer with image quality = 100 and the saved image (/tmp/crop.jpg) looks like

            unsigned char* crop_buffer1 = crop_interleaved(cropAddr, args.raw_width, args.raw_height, CHANNELS,
                                                                0, 0, 640, 640);
            unsigned long jpeg_size = 0;
            unsigned char* jpeg_buffer = NULL;
            struct jpeg_compress_struct jpeg_conf;
            set_jpeg_configuration(640, 640, CHANNELS, 100, &jpeg_conf);
            buffer_to_jpeg(crop_buffer1, &jpeg_conf, &jpeg_size, &jpeg_buffer);
            char file_name[32];
            snprintf(file_name, sizeof(char) * 32, "/tmp/crop.jpg");
            jpeg_to_file(file_name, jpeg_buffer, jpeg_size);
 
![crop](https://user-images.githubusercontent.com/27479245/209511892-0a5df378-83ba-42d4-af52-f6f904f892e0.jpg)
Is it normal output or is there any way to save a good image?
**Environment:**
Axis device model: P1375-E
Axis device firmware version: [Axis OS 10.08.1]
Stack trace or logs: [e.g. Axis device system logs]
OS and version: [ Ubuntu 20.04.03 LTS]

**Comment by pataxis** [comment]
Hi @amararjun , thanks for your question, we'll get back to you shortly.

**Comment by Corallo** [comment, marked as answer]
Hello @amararjun 

I tried to run the object detector example, and it seems to me that the output is correctly shown in the images.
Besides changing the resolution, did you modify something else from the original example?

The image that you get is not blurred, but some kind of offset shearing. Where does your `cropAddr` come from?

**Reply by AmarRobo** [reply]
Thank you for the response.

I tried running the sample program [object-detection] as it is. I changed chip from 4 to 2 in manifest.json; rest same as the original.
crop Addr have three appearances in the program:

first:
    if (!createAndMapTmpFile(CROP_FILE_PATTERN,
                             args.raw_width * args.raw_height * CHANNELS,
                             &cropAddr, &cropFd)) {
        goto end;
    }
second:
convertU8yuvToRGBlibYuv(args.raw_width, args.raw_height, nv12Data_hq, (uint8_t*) cropAddr);
third:
unsigned char* crop_buffer = crop_interleaved(cropAddr, args.raw_width, args.raw_height, CHANNELS,
                                                                  crop_x, crop_y, crop_w, crop_h);
This time too a  1925 X 919 image was saved as below:
![detection_0](https://user-images.githubusercontent.com/27479245/212244996-7d3b5d2a-1f45-4950-8446-e8e957e04bd1.jpg)

**Reply by Corallo** [reply]
I tried again on your model of camera, I can't replicate the issue that you have.

- Are you using a P1375-E? Make sure you use the [object-detection-example](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/object-detection) and not the [object-detection-cv25](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/main/object-detection-cv25)
- Update your firmware to a version >= 11.1.
- (extra) If you want to build the example to run for CPU, you can add --build-arg arch=cpu to the docker build command.

Check the first two points, maybe a firmware update can do the trick :)

**Reply by AmarRobo** [reply]
Thank you again for the reply.

I am using P1375-E. I have updated to the latest firmware 11.2. I use arch = cpu as there is no edgeTPU. After running the object-detection-example, when I copy .jpg file from /var/tmp folder using scp command,I am getting the similar offsetted image
![detection_0](https://user-images.githubusercontent.com/27479245/214244007-76eeb690-b11f-46d5-9e78-cf626a31da14.jpg)

I tried [opencv example](url)
.

**Reply by AmarRobo** [reply]
Thank you so much for investigating further!

I am using P1375-E and following the object detection example. I updated to latest firmware 11.2 with chip= cpu , the camera does not have edgetpu,I guess. Again, using scp command to copy from /var/tmp folder gives similar offset-ed image.

![detection_0](https://user-images.githubusercontent.com/27479245/214256271-d2dad894-3261-4ae2-813a-f2bdb77997fc.jpg)

I tried the [opencv example](https://github.com/AxisCommunications/acap-native-sdk-examples/tree/1d33b45c5ea557d4f8a5e217e00d5b23c2100765/using-opencv), I used imwrite function inside the example.cpp to save an image. The saved file too, after scp command gives similar image. I am using linux OS.

![test](https://user-images.githubusercontent.com/27479245/214256092-99fb4869-b562-4270-9ffe-9a07f141cd7d.jpg)

I am not sure where I am making the mistake or what causes this strange images.

**Reply by Corallo** [reply]
Can you show the code snipped that you used to save the image in the opencv-example?

**Reply by AmarRobo** [reply]
I added imgcodecs module and used imwrite function to save the image to tmp folder.
![code snippet](https://user-images.githubusercontent.com/27479245/214722114-49b3f328-859c-44ec-922b-efbc3900b838.png)

**Reply by Corallo** [reply]
Hi. I tried the same with an AXIS P1375. Firmware version 11.0.8
Editing line 12 of the Makefile with
`LDLIBS += -lm -lopencv_video -lopencv_imgproc -lopencv_imgcodecs -lopencv_core -lpthread`
and line 49 of the Dockerfile:
` -D BUILD_LIST=core,imgproc,video,imgcodecs \`
and adding to line 97 of the example.cpp
`imwrite("/tmp/img_bgr.jpg", bgr_mat);`

I get a normal image.
![img_bgr](https://user-images.githubusercontent.com/28478841/214800368-edfb73b8-aa6e-4b43-a841-5bbb4383e49a.jpg)

It is hard to pinpoint the issue with your camera, as it is not replicable. We can only advise you to factory default it, and then reinstall the latest firmware.
Verify that you are on the latest branch of the repository, (you can check that your docker file has `ARG VERSION=1.6`).
and also that connecting to the camera the video stream in the GUI is showed correctly

**Reply by AmarRobo** [reply]
Thank you for confirming that the program runs perfectly on AXIS P1375.
As for the camera I am working with,the video stream in the GUI is normal and I did exactly the same as you have mentioned, for running the opencv program.
I will try resetting the camera and reinstalling the firmware. Hopefully, I can get it correctly this time.
Thank you

**Reply by AmarRobo** [reply]
@Corallo 
Hi,
The problem with my image was that I had rotated the optic unit of the camera by around 150 degrees. So, I guess the width and height of the program and the data were not matching.
It got solved when I put the optic unit in the default position.

Thank you for the support!

**Reply by Corallo** [reply]
Great, thanks for letting us know!
